State Management Improvements

Pyxus opened the pull request:
The goal with these changes was to provide better support for nested state machines in addition to eliminating the need for the `CombatStateMachine`. I felt it needlessly introduced new concepts in the form of 'situations' which to some extent restricted how users could structure their state machines. Now there is only 1 state machine class which provides processing to the root and all its sub-states, and an optional advancer that can attempt to advance any state machine by feeding it buffered inputs.

Added:
- `BufferedInputAdvancer`: Contains input buffering functionality previously belonging to the `CombatStateMachine`. This is a node which attempts to manually advance a target state machine using inputs buffered in it. *I'd like any feedback on this class concept or its name.*
- Give `CompositeState`'s, formerly `RootState`, builder `transition()` method a 'transition' parameter. This allows users to pass custom transition objects into the system.

Changed:
- `goto()`, `has_state()`, and `get_state()` now accept state paths allowing you to more easily interact with sub-states.
- When using `goto(to_state_path)` the new behavior is to exit up each state leading to the most common ancestor of the current state and target state, and then follow along the path to the target starting from the most common ancestor state entering each state along the way.
- Default enter behavior of `CompositeState` is to set their current state to the start state. Additionally, sub-states are now entered if the current state is a `CompositeState`.
- Switched to using callables for conditions: https://github.com/Pyxus/fray/pull/70#issuecomment-1712960459

Renamed: 
- `RootState` -> `CompositeState`: This is to reflect the fact that this class contains a collection of sub-states including other composite states. I felt `RootState` didn't make that immediately obvious and may have implied it could only serve as the root of a state machine node.

Removed:
- `CombatStateMachine`: The combat state machine provided 2 features. A way of buffering inputs fed to it self, and a way of organizing root states into 'situations'. The `BufferedInputAdvancer` now provides the first functionality, and the improvements to using nested states now allows for sub states to take the place of situations. As pointed out by user it really was just a fancy name for a state machine that supported buffered inputs, I think this new approach is more flexible.

What's Left:
- The redesigned system requires testing and bug fixing for issues that appear
- Some signals need to be rethought. Namely `transitioned`
- `travel` method needs testing and potentially a redesign.
- Update documentation (Current plan is to do a full documentation overhaul soon enough)

Fixes: #64 
Fixes: #66

Remi123 commented on src/state_mgmt/state/composite_state.gd:
```diff
@@ -682,13 +765,13 @@ class Builder:
 	## Creates a new global transtion to the specified state. 
-	func transition_global(to: StringName, config: Dictionary = {}) -> Builder:
+	func transition_global(to: StringName, config: Dictionary = {}, sm_transition := FrayStateMachineTransition.new()) -> Builder:
 		var tr := _create_global_transition(to, FrayStateMachineTransition.new())
```
add sm_transition to the _create_global_transition argument

Remi123 commented:
Ok this is a big refactor. 

I'll state that I didn't read the new changes in the code yet, so I'll just comment on the description.

- Removed CombatStateMachine to BufferedInputAdvancer.

In theory this is a good idea, since there will always be some sort of combination of input or behavior that will fall outside of what the interface of CSM can help. I was already making some custom logic for my is_held vs is_tapped logic so if you provide a BufferInputAdvancer to define those "workaround" it can help and I'm in favor. However I didn't read the implementation so I don't know how easy it is. Is_tapped vs Is_held is for me a good test case. 
The naming could be better. I was thinking of my setup as a filter, because I only send interesting inputs (aka not `is_echo`) to the state machine, even buffering `is_tapped` even though it doesn't exist in my mapping. FrayInputFilter isn't really better.
It's basically the missing link between the input and state library.

- Removed CombatStateMachine to CompositeState with substate.

The `situations` technically were just a substate, and the lack of interaction between them is what make them difficult. I had to check the transition signal from my FrayRootState to properly manage my `change_situation` calls. I did find CSM very useful because of the reset that was occurring when changing situation, which returned the situation to the start state. This is useful for combos, or with generic state like "locomotion" or "falling" with check if you are on the ground or not. But in general it was easier to have everything in a big situation.

- Give CompositeState's, formerly RootState, builder transition() method a 'transition' parameter.

This is always nice to let users define their own object, but I didn't play enough with Transitions to know when it's a good idea apart from the obvious button and sequence. Maybe a FrayTimedTransition would be nice to add. I do look for a solution for a future problem where I need to communicate to my next state a value that I would calculate inside the transition.

Speaking of user-defined objects, I find conditions clunky to work with. I have to `get_current_situation().set_condition("in_air",not characterbody.is_on_floor())` on every frame. Is it possible to make condition extends Callable and rework the logic to evaluate if the name is the same and if it evaluate to true? Instead of a `Dictionary<String,bool>`, it's now a `Dictionary<String,Condition>`. It's more expensive but now the conditions can be used by multiple CompositeStates and only called when needed. I do like functional programming so I may be biased.

- Goto and CompositeState.

Goto behavior needs to be very well documented and tested. I think it is still useful to offer the possibility to teleport to the target_state, calling exit on the current one. I may be old school, but the name `goto` is reserved in my head to anything that just directly go to a target. This is more like a hierarchy_travel.

CompositeState default to go to start is great but I'm sure someone would want to keep the current substate on one node. Maybe upon exiting, set the start value of the CompositeState to the current substate if not resetable ? In any case, I need to test this onto my controller.

- Initial feedback.

Not sure if I helped 😄 or if I'm clear enough. But you are doing a great job and I do like the direction this is going. I'll keep you updated, I'm just a little bit busy.

Pyxus commented:
> I was already making some custom logic for my is_held vs is_tapped logic so if you provide a BufferInputAdvancer to define those "workaround" it can help and I'm in favor.

Could you help me better understand what you're trying to accomplish? Assuming you want to set up transitions that occur when a button "is_held" or "is_tapped". One of these is already possible. "is_held" would be a button transition with a min_time_held set.  "is_tapped" should just require me to add a max_time_held constraint.

You also mention your setup being similar to a filter so I may just be misunderstanding your goals.

> Is it possible to make condition extends Callable and rework the logic to evaluate if the name is the same and if it evaluate to true

Good idea, i'll look into this. The current condition setup was implemented long before Callables existed, and I wasn't a fan of Godot 3's FuncRefs at the time.

> I may be old school, but the name `goto` is reserved in my head to anything that just directly go to a target. This is more like a hierarchy_travel.

After thinking about it I believe it would be better to rename this as it is more similar to travel than the previous goto. It just travels vertically instead of horizontally. And `goto` likely would be more widely recognized as "go directly to". Especially for those coming from languages with labels like C. That being said i'll re-add `goto` with this behavior.

> CompositeState default to go to start is great but I'm sure someone would want to keep the current substate on one node.

Good point. This could be resolved by adding a `is_persistent` flag to CompositeState. This way users can decide when building if they'd like their CompositeStates to reset the current state to start on enter 

> It's more expensive but now the conditions can be used by multiple CompositeStates and only called when needed. I do like functional programming so I may be biased.

I think the greater ease of use trumps the added expense so I'm all for this idea.

> Not sure if I helped 😄 or if I'm clear enough. But you are doing a great job and I do like the direction this is going. I'll keep you updated, I'm just a little bit busy.

Thank you! I appreciate you taking the time to leave some feedback

Pyxus commented:
## Update: String to Callable mappings are now used to represent conditions

When building composite states the new `register_conditions` method can be used to make the system recognize a desired condition.

Example:
This example represents an attack that can only be performed after another attack lands.
```
FrayCompositeState.builder()
   .register_conditions({
      on_hit = func() return true
   })
   .transition_button("idle", "attack1")
   .transition_button("attack1", "attack2", {input="btn_punch", prereqs=["on_hit"]}
   .build()
```

It is also possible to check for the inverted condition using the '!' prefix like so:
```
transition("a", "b", {prereqs=["!on_hit"]})
```
Lastly all states within a hierarchy reference conditions registered on the root. So conditions only need to be declared once to be used everywhere.

Remi123 commented:
I'm very eager to try those new callable-based conditions. It give new opportunities especially with raycastting during specific states.

For `goto` names is ultimately not important. as long as it's documented, but it's important that the user know that there is options.  It's your library after all.

About my issue, it's probably me trying to fit the is_tapped behavior inside the combatstatemachine and probably making a small error that made the thing a little bit more difficult than it should be. You already had an idea on how you would have done. I've tried to implement it with a max_time_held, but I probably make a mistake somewhere.

My solution was to filter the FrayInputEvent to remove mostly is_echo event, then in the case of my tapped button is released I check the time held and if below a threshold I send a custom button to CSM named "X_tapped". The is_held behavior is part of one of the state to run. So technically the is_held is not part of the CSM input dispatching, Basically there is a few lines of code in the FrayInput.input_detected signal ( or whatever the name is). Just document somewhere that the CSM doesn't need echo event.

I'm just trying to explain my thought process and my errors. I think it's valuable to you as I'm trying to do an platformer/action game where the conditions a more external ( is on floor, has hit a wall, etc).

Once my 2 years old get to sleep tomorrow I'll try to test this branch.

Pyxus commented:
> I've tried to implement it with a max_time_held, but I probably make a mistake somewhere.

Just for clarification `max_time_held` is *currently* not a property that InputTransition has. However, it does have `mind_time_held`. I believe that `min_time_held` can capture the case where you wish for a transition to occur 'on button held'. And I was suggesting that a `max_time_held` could be included to capture the case where you wish for a transition to occur `on button tapped`.

So basically, in the context of state machine transitions...
is_held = pressed and time_held > min_time_held
is_tapped = released and time_held < max_time_held

I am planning to implement this now that it has been brought to my attention since I believe it could be useful. However, I was uncertain if this would address the limitations you've encountered. Correct me if I'm wrong, but I think it does 🤔.

Pyxus commented:
# Notable Updates

Added:
- Added `max_time_held` property to input transition
-  Added `_input_impl`, `_unhandled_input_impl`, and `_init_state_impl` virtual methods to `State` class.

Renamed:
- `CompositeState` to `CompoundState`. It is minor but I feel this name better reflects the purpose of the class. It also prevents confusion between `CompositeState` and `CompositeInput`. This should be the final name.

Changed:
- Input transitions now use milliseconds in the form of integers for timings. In general, I want to make it a standard to present all input-related times in Fray as integer milliseconds. I believe most input-related time checks only occur within the span of seconds so I see no reason to use seconds or floats.

What's Left:
- Testing, especially `travel` and `goto`
- Update documentation (Plan is still to do an overhaul)
- Rethink `transitioned` signal. Currently, I'm leaning toward it showing the full path. The 'ideal' behavior depends on what is most useful to report which will require me to plan around with the tool more until I recognize some pain-points. However, I'm also opened to suggestions.

Remi123 commented on src/state_mgmt/state/compound_state.gd:
```diff
@@ -79,17 +79,37 @@ var _current_state: StringName
+		cur_state._input_impl(event)
+
+
+func _unhandled_input_impl(event: InputEvent) -> void:
+	var cur_state := get_current_state()
+	if cur_state is FrayCompoundState:
+		cur_state._input_impl(event)
```
copy paste error : should be calling `cur_state._unhandled_input_impl'

Remi123 commented on src/state_mgmt/state_machine.gd:
```diff
@@ -3,129 +3,136 @@ class_name FrayStateMachine
-		root.current_state = value
+		root = value
+		root._enter_impl({})
+		root._init_state_impl()
```
_init_state_impl shouldn't be called in the root setter. Only when building the states in add_state.

Remi123 replied:
My mistake, the root setter is for applying the CompoundState to the StateMachine.

Remi123 commented:
Didn't have time to test, will do it soon. Here is some of my comments

>     * Added `max_time_held` property to input transition

Thank you

>     * Added `_input_impl`, `_unhandled_input_impl`, and `_init_state_impl` virtual methods to `State` class.
>     * Added ability to set default state and transition in the compound state builder.  These defaults affect every object instantiated after their call so call order is important. This was added to make building systems with custom states less tedious. However, the downside is these states can't have constructors.

_input and _unhandled_input are good addition. My issue with `_init_state_impl` is that `_init` is a special function name that gdscript interpret as a constructor when calling `MyClass.new(args...)` so I'm not sure if it's possible to mix your API. 
Having a default state is not something that I care too much about so.
 

>     * Input transitions now use milliseconds in the form of integers for timings. In general, I want to make it a standard to present all input-related times in Fray as integer milliseconds. I believe most input-related time checks only occur within the span of seconds so I see no reason to use seconds or floats.

The issues is that it's more difficult to distinguish if an integer is a millisecond or a framecount. At least with float in second it's 0.016 for 1 frame, so it's visually different instead of 16 and 1. I'm really have a preference for seconds in float since it's standard but I'm not someone that goes into framedata of my characters in fighting game. At least we need an helper function that convert seconds to frames, ms to frames and vice versa. 


>     * Testing, especially `travel` and `goto`
> 
>     * Update documentation (Plan is still to do an overhaul)

For travel and goto, it's both test and document the behavior with hierarchy. I will probably help with the documentation.


>     * Rethink `transitioned` signal. Currently, I'm leaning toward it showing the full path. The 'ideal' behavior depends on what is most useful to report which will require me to plan around with the tool more until I recognize some pain-points. However, I'm also opened to suggestions.

I think that this will heavily depend on the `match` keyword of godot. I can only speak for myself but my logic is either check if the transition goes to a specific state, or since we get hierarchy if we enter a parent state.

``` GDScript
func on_transitioned(from:StringName,to:StringName):
  match [from,to]:
    case ["A",_]:
      prints("A to anything")
    case [_,"A/a"]:
      prints("Anything to A/a")
```
It's a powerful tool. I think the best for now would be to leave it full path for now, provide one or two functions to get the hierarchy of the state by it's path, and let the user do its string manipulation however it want.  The String class have some function to manipulation filepath that could be useful like `String.get_basename()`

``` GDScript
func on_transitioned(from:StringName,to:StringName):
  var folder = str(from).get_base_dir()
  match [folder ,to]: # matching the parent of the from state to destination state in full path
    case ["A","A/a"]:
      prints("Any state in A to A/a in particular")
    case [_,"A/a"]:
      prints("Anything to A/a")
```

Remi123 commented:
My first impression is good.

One strange bug I have is that I have a custom fraystate for one of the state, but only the enter_impl is called. Exit_impl isn't called. 
Didn't tried to find the problem yet.

I found that adding substate quite easy if i use two builder and add the substate. 

The default state being gdscript class isn't as versatile as I hoped. 

Conditions being callable is a win. So easy.

I'll share the test case soon. But it's a very barebone parent-child hierarchical state machine

Remi123 commented:
Found the issue about the exit function not being called on my custom fraystate. I changed it to extends FrayCompoundState instead of FrayState and the exit function is now called. I don't know how fraystate are supposed to interact within FrayCompoundState, but it's something to add to the docs. The issues remains that classes that extends FrayState doesn't get called their exit function.

For adding substate, here how I did it : 

```GDScript
extends CharacterBody3D

@onready var sm: FrayStateMachine = $FrayStateMachine
@onready var buffer: FrayBufferedInputAdvancer = $FrayBufferedInputAdvancer

func _unhandled_input(event: InputEvent) -> void:
	if event.is_action_pressed('ui_right'):
		buffer.buffer_button('btn_punch')
	if event.is_action_pressed('ui_left'):
		buffer.buffer_button('btn_kick')

var combat : FrayCompoundState
var loco : FrayCompoundState

func _ready() -> void:
	combat = (FrayCompoundState.builder()
		.start_at('start')
		.add_state("punch",DebugTrace.new("Punch")) # DebugTrace is printing when entering and exiting states
		.add_state("kick",DebugTrace.new("Kick")) # kick attack, another entry point

		.transition('start','punch',{auto_advance=true}) # Default attack is the punch
		.transition_button("punch", "combo", {input="btn_punch", prereqs=["on_hit"]}) #can combo on hit


		.tag_multi(['punch',"combo","kick"],['finish']) # Tagged finish combo

		.tag('end',['to_end'])
		.add_rule('finish','to_end') # transition to end rule
		.transition_global('end',{auto_advance=true,prereqs="!on_hit"}) # auto end combo
		.end_at('end')

	).build()

	loco = (FrayCompoundState.builder()
		.register_conditions({
			on_hit = func(): return ..., # hit logic, not ready yet
			on_ground = func(): return is_on_floor(),
			in_air = func(): return not is_on_floor() and not is_on_wall(),
			on_wall =func():return is_on_wall()
		})
		.start_at('ground')
		.add_state('combat',combat) # add the compoundstate of combat to the combat state inside locomotion.
		.add_state('ground',DebugTrace.new('on ground'))
		.transition('air','ground',{prereqs=['on_ground'],auto_advance=true})
		.transition('ground','air',{prereqs=['in_air'],auto_advance=true})
		.transition_button('ground','combat',{input="btn_punch"}) # can enter combat state, auto_advance to punch
		.transition_button('ground','combat/kick',{input="btn_kick"}) # Kick
		.transition('combat','ground',{auto_advance=true,switch_mode=FrayStateMachineTransition.SwitchMode.AT_END}) # Return to ground when attack finish.

	).build()

	sm.root = loco # assign locomotion as the root
```

This is REALLY easy. Not sure if this is how you would do it but I'm already imagining a lot of functionality from this.

Ideally I would like to register the conditions in their substate that require them ( loco doesn't need on_hit, but I must register it there otherwise it isn't registered). Not a deal breaker.

Also, about the default state, the fact that it's a GDScript class doesn't allow me to configure it as I go, but I don't know how to improve it. For now I'm mainly using `add_state("state",Whatever.new(my_configuration_args))`

Remi123 commented:
Unless you have other changes that you want to make, I think this is a great refactoring. It work like a charm on my test cases and the conditions are easy to setup. Hierarchical states solved my previous problem with multiple situations in CombatStateMachine,

About what's left:

`transitioned` should be full path. There is no way to imagine all possibles scenario where a user want to react to the change from one substate to another. String parsing goes a long way for filtering states transitions. 

About `FrayCompoundState`renaming , maybe you can switch the ordering to help where each class belongs ? Something like `FrayStateCompound` and `FrayInputComposite` would help with the auto-complete. Obviously make the branch more than a redesign for state but I think it's worth it. You decide.

I think the only thing missing is some helpers for transition using a signal. Animations have a lot of points of interest like `on_hit` and I'm not sure what is the canonical way of interacting with it using `FrayStateMachine`. For now it's either something that I manage within the `FrayState`, or whenever the on_hit signal is trigger I send a `BufferedInput` through the `FrayBufferedInputAdvancer `, which is bad because it's buffered in some cases. It would be nice to connect Timer's Signal with this.

I'll help with the documentation soon

Kuenaimaku commented:
Hey all, I've been messing around with this PR in a pretty generic 3d environment and for the most part, I think that this refactor is very well thought out! The builder syntax is very descriptive, and the `register_conditions` 

`transitioned` being a full path makes sense to me, as with the current design there's no limit to the number of `FrayCompoundState`s that could make up a StateMachine.

I'm in agreement with Remi123 on helpers for transitions. Outside of rolling my own `FrayState` I don't see a great way of making it so the animation can control when a state can be considered "finished processing", which makes sense given how the intent is for these 3 modules to act independently of each other. 

What I'm aiming to do kinda already goes against keeping these 3 modules separate, as I'd like to annotate my animations with all these properties (what FrayState is linked to a HitState? At what point can this animation be considered finished processing to return back to the idle state? at what point can a player input a button to 'cancel' into another action?) rather than having it mostly.

In my case, I'm trying to replicate a JRPG's action combat system, in which a single character has 3 attacks that can be comboed from 1, to 2, to 3. These animations have a very long follow-through (pretty much putting them in an idle state) in which the player can delay the next part of the string, in order to have other characters come in and do parts of their combo. In attempting to emulate this, I had the animation player controlling both the `FrayStateMachine.Active` property as well as the `FrayBufferedInputAdvancer.Paused` property. Controlling the `FrayStateMachine` like this doesn't seem right, since if they get hit, I would want them to swap to a hurt state, and if that's inactive, then that will never process.

For example, this is the test player's script:

```GDScript
extends CharacterBody3D

@onready var sprite = $Sprite3D
@onready var animation_player = $AnimationPlayer
@onready var state_machine: FrayStateMachine = $FrayStateMachine
@onready var input: FrayBufferedInputAdvancer = $FrayBufferedInputAdvancer

var state_container: FrayCompoundState

var target_velocity = Vector3.ZERO
var direction = Vector3.ZERO
var button_input_name = null

func _ready() -> void:
	_setup_state_machine()	

func _unhandled_input(event: InputEvent) -> void:
	if event.is_action_pressed(button_input_name):
		input.buffer_button('attack')
	if event.is_action_pressed("battle_shoulder_right"):
		input.buffer_button('special')
		
func _setup_state_machine() -> void:
	state_container = ( FrayCompoundState.builder()
		.register_conditions({
			on_hit = func(): return true,
		})
		.transition('start', 'idle')
		.transition_button('idle', 'neutral_attack_1', { input='attack' })
		.transition_button('neutral_attack_1', 'neutral_attack_2', { input='attack' })
		.transition_button_global('special_attack', { input='special' })
		.tag_multi(['idle'], ['neutral'])
		.tag_multi(['neutral_attack_1', 'neutral_attack_2',], ['normal'])
		.tag_multi(['special_attack'], ['special'])
		.add_rule('neutral', 'special')
		.add_rule('normal', 'special')
		.add_rule('normal', 'neutral')
		.add_rule('special', 'neutral')
		.start_at('start')
	).build()
	state_machine.root = state_container
	animation_player.play('start')
	
func _on_fray_state_machine_state_changed(from, to):
	print("state change from: " + from + " to: " + to)
	if(anim_player.get_animation_list().has(to)):
		anim_player.stop()
		anim_player.play(to)
```
[and an image link that hopefully helps explain my current setup.](https://i.imgur.com/2aliaF5.png)

I also tried to make it so that every animation would return to idle by setting them up as a default transition ( something like `.transition('neutral_attack_1', 'idle')` ) with the intent that at the end of the attack animation, they would return to an idle state, but that just made it so the attack state would immediately swap to the idle state, which makes sense given how `FrayState` currently works.

One final note, an example project to showcase a typical setup similar to the example gif in the project's README would be greatly appreciated, as trying to use the documentation was a bit painful.

Remi123 commented:
@Kuenaimaku 

I wouldn't really control the state machine active property, as it add complexity when it's active or not. I would add a lot more `prereqs` and conditions to manage that.

```GDScript
@export var can_transition :bool = false

...

root_machine.register_conditions(
  {can_transition = func() return self.can_transition,
  }
)
.transition_button('idle', 'neutral_attack_1', { input='attack' })
.transition_button('neutral_attack_1', 'neutral_attack_2', { input='attack', prereqs=[can_transition,on_hit] })
```

Then modify the `can_transition` property in the animations.

Remi123 commented:
I've found a small bug in the collision library. 

```GDScript
# hit_state_3d.gd
## Sets whether the hitbox at the given [kbd]index[/kbd] is active or not.
func set_hitbox_active(index: int, is_active: bool) -> void:
	var flag = 1 << index

	#if flag > active_hitboxes: 
		#push_warning("Index out of bounds.")
		#return
        # <-- Removed this warning since once all hit_box are disable with active_hitboxes = 0,
        #       it prevent me from re-activate them.

	if is_active:
		active_hitboxes |= flag
	else:
		active_hitboxes &= ~flag # <-- missing the ~
```

Pyxus commented:
> My issue with `_init_state_impl` is that `_init` is a special function name that gdscript interpret as a constructor when calling `MyClass.new(args...)` so I'm not sure if it's possible to mix your API.

This should be fine, _init is a special function but Godot doesn't treat _init_XYZ as anything special. That being said I settled on the name _ready_impl since I think that better captures that the virtual method is used for.

> The issues is that it's more difficult to distinguish if an integer is a millisecond or a framecount. At least with float in second it's 0.016 for 1 frame, so it's visually different instead of 16 and 1. I'm really have a preference for seconds in float since it's standard but I'm not someone that goes into framedata of my characters in fighting game. At least we need an helper function that convert seconds to frames, ms to frames and vice versa.

A few other devs I know agreed that unless specifically told they assume time is in seconds. Additionally I think your argument of ms and frame count being confused holds some weight. I've gone back to having the time in seconds, and added static methods under the `Fray` class_name to convert between seconds and ms. This also contains a frames_to_msec and vice-versa method if you weren't already aware.

Also going forward, to keep things consistent, all times in Fray should use seconds, and non-second times will be given an appropriate suffix.

> Found the issue about the exit function not being called on my custom fraystate. I changed it to extends FrayCompoundState instead of FrayState and the exit function is now called. I don't know how fraystate are supposed to interact within FrayCompoundState, but it's something to add to the docs. The issues remains that classes that extends FrayState doesn't get called their exit function.

This should be fixed.

> Ideally I would like to register the conditions in their substate that require them ( loco doesn't need on_hit, but I must register it there otherwise it isn't registered). Not a deal breaker.

Noted. What I will likely do is support global conditions (located on the root) and local conditions. Conditions will be checked locally first, and then globally so local conditions will shadow global conditions. I'll also add a character to explicitly check a global condition. Maybe: "$on_hit"?


> One final note, an example project to showcase a typical setup similar to the example gif in the project's README would be greatly appreciated, as trying to use the documentation was a bit painful.

Thanks for checking out the project! And yes the documentation needs a big overhaul. This PR will be merged within the next few days and after that my next priority is rewriting the documentation and including examples.

> I think the only thing missing is some helpers for transition using a signal. Animations have a lot of points of interest like `on_hit` and I'm not sure what is the canonical way of interacting with it using `FrayStateMachine`. For now it's either something that I manage within the `FrayState`, or whenever the on_hit signal is trigger I send a `BufferedInput` through the `FrayBufferedInputAdvancer `, which is bad because it's buffered in some cases. It would be nice to connect Timer's Signal with this.

> I'm in agreement with Remi123 on helpers for transitions. Outside of rolling my own `FrayState` I don't see a great way of making it so the animation can control when a state can be considered "finished processing", which makes sense given how the intent is for these 3 modules to act independently of each other.

I'll create a discussion soon enough for this so I can better understand what sort of solution might address this pain point. It won't be addressed in this PR.

> About `FrayCompoundState`renaming , maybe you can switch the ordering to help where each class belongs ? Something like `FrayStateCompound` and `FrayInputComposite` would help with the auto-complete. Obviously make the branch more than a redesign for state but I think it's worth it. You decide.

Would be great if Godot had support for namespaces... I'll give it some thought and tackle this in another PR if I decided to go forward.

Remi123 commented on src/state_mgmt/state/state.gd:
```diff
@@ -1,5 +1,5 @@
 class_name FrayState
-extends RefCounted
+extends Resource
```
FrayState becoming a Resource is interesting. Saving and referencing other state machine should be easier now with `Resource.load`. Great change

Pyxus replied:
Transitions will also become a Resource. I had meant to include the change in this commit but I just noticed while testing that it slipped my mind. Will be pushing a fix shortly.

Remi123 commented on src/state_mgmt/state/compound_state.gd:
```diff
@@ -713,10 +716,12 @@ func _goto(path: StringName, args: Dictionary = {}) -> void:
 		if state == common_active_ancestor:
 			break
+		
+		if state is FrayCompoundState:
+			state._current_state = ""
```
I'm sure we need to check if the check is persistent before setting the current state to empty.
```suggestion
		if state is FrayCompoundState and state.is_persistent:
			state._current_state = ""
```

Pyxus replied:
Good catch. Also, seeing this again made me realize I can just move this logic into compound_state.gd's exit method. That feels better than checking the state's type.

Remi123 reviewed (changes requested):
FrayState to Resource is a feature that need to be added to the doc !

Except my comment about persistence, the change are solid.

Remi123 commented:
[oct31st.webm](https://github.com/Pyxus/fray/assets/16803994/a371b5fd-ce98-4358-b467-9e0ecd78f136)

During the last weeks I've been experimenting with this branch and implemented everything in this video. Don't mind the lack of walking animation I'm working on a custom animationplayer and it's a separate project. Jumping, rolling and the ladder are all `fraycompoundstate` and I want to share my experience with implementing that.

1. 

First, I'd like to share that since it's more of an action-platformer, I didn't do much in terms of combos so I've learned to use less the builder and add my transitions manually. I think I was a little bit too much focused on the builder previously, which is great for combos and combat-related overview, but I was looking to encapsulate some behavior. This is where the newly named `_ready_impl()` function come into play and this is useful to know that all my states are properly setup before adding more transitions. 

```GDScript
# In JumpState.gd extends FrayCompoundState
func _ready_impl():
  set_condition(...) # Local condition are supported in this setup
  get_root().transition_button("ground","jump",setup) # Dependency on root having ground but I know it has it.
  add_state("anticipation") # Deal with internal states
  ...  
```
The only way I would improve this is to find a way to link object. My state modify the position of the `characterbody3d` a lot, so each of my state need to have a constructor where I pass all my info since they are Objects, not Node. Maybe the StateMachine Node could have a `blackboard` dictionary, and it pass the blackboard into the `_ready_impl' and it propagate? I'm just trying to reduce this boilerplate but it's not a big deal if there is no perfect solution.

2.

Second, seconds being the standard time reference is great. I have some plans that will need `frame_to_sec` so this function is appreciated.

3.

> I believe resource duplication only copies exported members
Caution about this stuff. Resource memory management is heavily dependant on `resource_local_to_scene` aka Make Unique. My rule of thumb is that if `local_to_scene` is false, next `duplicate()` return the reference to the object. If set to true, next `duplicate()` actually copy the object.

4.

`"$on_hit"` Allowing local conditions... I don't know. If I reread my comments, It's clear that I was still trying to understand how `compoundstate` works and I didn't try yet to not use the builder. Of course I wanted correct local conditions since I was cramming everything into one big `builder()...` statement. My ladder have a few local conditions and properly building the state using `_ready_impl` is more versatile. I won't insist on this feature since I clearly was still learning when I wrote this, but if you like the idea of overriding conditions and you can think of a usecase, it can be useful.

5.

I've implemented a relatively simple state designed to play an animation and await either the end of the anim, or a signal. In fact all "actions" in the video is using this state. I know you will create a discussion about this but here is my pseudo-code on this state so you can think about it:

``` GDScript
# The signal can_advance is set in the animation
var ended :bool = false
func _is_done_processing_impl() -> bool:
  return ended
func anim_finish(_anim_name = ""):
  ended = true
  get_root().advance()

func _enter_impl(args:Dictionary):
  ended = false
  if not statemachine.can_advance.is_connected(on_anim_finish):
    statemachine.can_advance.connect(anim_finish,CONNECT_ONE_SHOT)
  if not animationplayer.animation_finished.is_connected(on_anim_finish):
    animationplayer.animation_finished.connect(anim_finish,CONNECT_ONE_SHOT)
  animationplayer.play(args.anim_name)

func _exit_impl() -> void:
  ended = false
  # Make sure we disconnect signals.
  if statemachine.can_advance.is_connected(anim_finish):
    statemachine.can_advance.disconnect(anim_finish)
  if animationplayer.animation_finished.is_connected(anim_finish):
    animationplayer.animation_finished.disconnect(anim_finish)
```

Conclusion : 

I continue to think this is a great library, and this is a great redesign. Can't wait to pull those last changes.

Pyxus commented:
> The only way I would improve this is to find a way to link object. My state modify the position of the `characterbody3d` a lot, so each of my state need to have a constructor where I pass all my info since they are Objects, not Node. Maybe the StateMachine Node could have a `blackboard` dictionary, and it pass the blackboard into the `_ready_impl' and it propagate? I'm just trying to reduce this boilerplate but it's not a big deal if there is no perfect solution.

I've actually considered doing something like this for a while now. It shouldn't be a difficult inclusion, but I won't be addressing it in this PR. I'll start a discussion to share my thoughts on potential approaches before I proceed.

> `"$on_hit"` Allowing local conditions... I don't know. If I reread my comments, It's clear that I was still trying to understand how `compoundstate` works and I didn't try yet to not use the builder. Of course I wanted correct local conditions since I was cramming everything into one big `builder()...` statement. My ladder have a few local conditions and properly building the state using `_ready_impl` is more versatile. I won't insist on this feature since I clearly was still learning when I wrote this, but if you like the idea of overriding conditions and you can think of a usecase, it can be useful.

Oh I see. It was very easy to add support for local conditions so when I saw you mention your experience I went ahead and implemented it. Global conditions are really just the root's local conditions. I personally can't think of a use-case in this moment but I think i'll keep it as I do like the idea. I see the behavior as similar to a function scope var shadowing a class scope var, but the class var can still be accessed with 'self'. Also the distinction between global and local isn't that significant.

---

I'll be merging this PR tonight or tomorrow as i'm satisfied with the current state of the redesign. The remaining pain points / Quality of Life improvements will be dealt with in subsequent PRs. Thanks for your ongoing interest and support in this project! Your feedback and code snippets have been very helpful while working on this. It's just also interesting to see how other people make use of the library.

Pyxus reviewed (commented):
No obvious issues from my tests or final review. Looks good.

Pyxus commented:
> The only way I would improve this is to find a way to link object. My state modify the position of the `characterbody3d` a lot, so each of my state need to have a constructor where I pass all my info since they are Objects, not Node. Maybe the StateMachine Node could have a `blackboard` dictionary, and it pass the blackboard into the `_ready_impl' and it propagate? I'm just trying to reduce this boilerplate but it's not a big deal if there is no perfect solution.

I decided to share my thoughts in an issue rather than using discissions. Let me know what you think when you have time.
https://github.com/Pyxus/fray/issues/72

Pyxus commented:
> I think the only thing missing is some helpers for transition using a signal. Animations have a lot of points of interest like `on_hit` and I'm not sure what is the canonical way of interacting with it using `FrayStateMachine`. For now it's either something that I manage within the `FrayState`, or whenever the on_hit signal is trigger I send a `BufferedInput` through the `FrayBufferedInputAdvancer `, which is bad because it's buffered in some cases. It would be nice to connect Timer's Signal with this.



> I'm in agreement with Remi123 on helpers for transitions. Outside of rolling my own `FrayState` I don't see a great way of making it so the animation can control when a state can be considered "finished processing", which makes sense given how the intent is for these 3 modules to act independently of each other.

This is being discussed here: https://github.com/Pyxus/fray/discussions/73
